File: python/src/spreen_wiki/unknown_wiki_count_list_exporter.py
import os
from typing import Any

from .application import Application

DEFAULT_OUTPUT_FILENAME = 'unknown_wiki_count_list_by_namespace.txt'


class UnknownWikiCountListExporter(Application):
    """Writes a sorted text report of how many wikis exist under each of
    the configured "unknown owner/category" namespaces."""

    _cached_counts: list[str]

    def __init__(self, output: str | None = None, **options: Any) -> None:
        super().__init__(**options)
        self._path_to_export = os.path.join(
            self._base_path, output or DEFAULT_OUTPUT_FILENAME)

    # private

    def _run(self) -> tuple[list[str], str]:
        counts = self._count_list_by_namespace()
        with open(self._path_to_export, 'wb') as f:
            f.write(('\n'.join(counts) + '\n').encode())
        return counts, self._path_to_export

    def _namespace_list(self) -> list[str]:
        return self._config.unknown_namespaces(self._group_by,
                                               self._language)

    def _missing_count_list_by_namespace(self) -> list[str]:
        plain_keys = set(self._plain_wiki_maps().keys())
        return [
            f'{namespace}: 0'
            for namespace in self._namespace_list()
            if namespace not in plain_keys
        ]

    def _count_list_by_namespace(self) -> list[str]:
        if hasattr(self, '_cached_counts'):
            return self._cached_counts
        plain = self._plain_wiki_maps()
        counts = [
            f'{namespace}: {len(plain[namespace])}'
            for namespace in self._namespace_list()
            if namespace in plain
        ]
        self._cached_counts = sorted(
            counts + self._missing_count_list_by_namespace())
        return self._cached_counts

File: RubyGem/lib/spreen_wiki/unknown_wiki_count_list_exporter.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'application'

module SpreenWiki
  # Writes a sorted text report of how many wikis exist under each of the
  # configured "unknown owner/category" namespaces.
  class UnknownWikiCountListExporter < Application
    DEFAULT_OUTPUT_FILENAME = 'unknown_wiki_count_list_by_namespace.txt'

    # @rbs base_path: String
    # @rbs group_by: String
    # @rbs language: String
    # @rbs home_overflow: (bool | String)
    # @rbs output: String?
    # @rbs **options: untyped
    # @rbs return: void
    def initialize(base_path: Dir.pwd, group_by: 'Owner', language: 'English', home_overflow: 'false',
                   output: nil, **)
      super(base_path:, group_by:, language:, home_overflow:, **)
      @path_to_export = resolve_output_path(output || DEFAULT_OUTPUT_FILENAME)
    end

    # @rbs return: [Array[String], String]
    def run
      File.open(path_to_export, 'wb') { |f| f.puts(count_list_by_namespace) }
      [count_list_by_namespace, path_to_export]
    end

    private

    attr_reader :path_to_export #: String

    # @rbs filename: String
    # @rbs return: String
    def resolve_output_path(filename)
      File.absolute_path?(filename) ? filename : File.join(base_path, filename)
    end

    # @rbs return: Array[String]
    def namespace_list
      config.unknown_namespaces(group_by, language)
    end

    # @rbs return: Array[String]
    def missing_count_list_by_namespace
      (namespace_list - plain_wiki_maps.keys).map { |namespace| "#{namespace}: 0" }
    end

    # @rbs return: Array[String]
    def count_list_by_namespace
      @count_list_by_namespace ||= begin
        counts = plain_wiki_maps.slice(*namespace_list).map { |namespace, wikis| "#{namespace}: #{wikis.length}" }
        (counts + missing_count_list_by_namespace).sort
      end
    end
  end
end
